### PR TITLE
Making ImagePullPolicy configurable for helper pod

### PR DESCRIPTION
**Pull request by ipsita2192**
Adding variable LIBImagePullPolicy to make ImagePullPolicy configurable for helper pod..

### CODE REVIEW

File: chaoslib/litmus/node-memory-hog/lib/node-memory-hog.go [hunk 1]
```diff
@@ -257,7 +257,7 @@ func CreateHelperPod(experimentsDetails *experimentTypes.ExperimentDetails, appN
 				{
 					Name:            experimentsDetails.ExperimentName,
 					Image:           experimentsDetails.LIBImage,
-					ImagePullPolicy: apiv1.PullAlways,
+					ImagePullPolicy: apiv1.PullPolicy(experimentsDetails.LIBImagePullPolicy),
```
Comment (ispeakc0de) [inline chaoslib/litmus/node-memory-hog/lib/node-memory-hog.go:260]: Would you like to add the same changes for [pumba lib](https://github.com/litmuschaos/litmus-go/tree/master/chaoslib/pumba) also?

Comment (SamarSidharth) [reply chaoslib/litmus/node-memory-hog/lib/node-memory-hog.go:260]: Yes Shubham sure we can do it

Comment (SamarSidharth) [reply chaoslib/litmus/node-memory-hog/lib/node-memory-hog.go:260]: Done the changes for pumba lib aswell

Comment (SamarSidharth) [reply chaoslib/litmus/node-memory-hog/lib/node-memory-hog.go:260]: We have done the changes for Pumba lib aswell. Kindly check

**Review by ispeakc0de — APPROVED** [review]

**Comment by ksatchit** [comment]
Thanks @ipsita2192 & @SamarSidharth - for creating this PR!! Hoping this is the start of my contributions !!

**Review by ksatchit — APPROVED** [review]

**Comment by arunsriniv** [comment]
Hi, Is this fix available in litmuschaos 1.13.2 ? I am facing this issue on litmus 1.13.2

**Comment by SamarSidharth** [comment]
Hi @arunsriniv, This is fixed you need to update it in the experiment file as below.

experiment.yaml
    image: external/go-runner:1.13.2
    imagePullPolicy: IfNotPresent
   
   env:
     - name: LIB_IMAGE
       value: 'external/go-runner:1.13.2'

     - name: LIB_IMAGE_PULL_POLICY
        value: 'IfNotPresent'

**Comment by arunsriniv** [comment]
Thanks @SamarSidharth. I tried this. But still its trying to pull image from Internet. And its failing.

**Experiment Yaml:**
image: "litmuschaos/go-runner:1.13.2"
    imagePullPolicy: IfNotPresent
    args:
    - -c
    - ./experiments -name pod-cpu-hog
    command:
    - /bin/bash
    env:
    - name: TOTAL_CHAOS_DURATION
      value: '60'  
  - name: LIB_IMAGE_PULL_POLICY
      value: 'IfNotPresent'

Describe pod on Helper pod:

    Command:
      /bin/bash
    Args:
      -c
      ./experiments -name pod-cpu-hog
    State:          Waiting
      Reason:       ImagePullBackOff
    Ready:          False
    Restart Count:  0
    Environment:
      CPU_CORES:              1
      SOCKET_PATH:            /var/run/docker.sock
      TARGET_PODS:
      CHAOS_NAMESPACE:        avc
      ANNOTATION_KEY:         litmuschaos.io/chaos
      AUXILIARY_APPINFO:
      LIB_IMAGE:              litmuschaos/go-runner:1.13.2
      SEQUENCE:               parallel
      **LIB_IMAGE_PULL_POLICY:  Always**

**Comment by SamarSidharth** [comment]
@arunsriniv can you share your experiment file

**Comment by arunsriniv** [comment]
> @arunsriniv can you share your experiment file

Pls find it attached as txt file

[1.13.2_generic_experiments.txt](https://github.com/litmuschaos/litmus-go/files/6466144/1.13.2_generic_experiments.txt)

**Comment by arunsriniv** [comment]
@SamarSidharth - any update? Anything wrong with the experiments file. I tried the same with 1.13.4 also. But facing same issue. Not sure where it is going wrong.

**Comment by ispeakc0de** [comment]
Hi @arunsriniv 

Can you please share the chaosengine manifest?

**Comment by arunsriniv** [comment]
apiVersion: litmuschaos.io/v1alpha1
kind: ChaosEngine
metadata:
  name: avc-platform-gateway-chaos
  namespace: avc
spec:
  # It can be true/false
  annotationCheck: 'false'
  # It can be active/stop
  engineState: 'active'
  appinfo:
    appns: 'avc'
    applabel: 'app=avc-platform-gateway'
    appkind: 'deployment'
  chaosServiceAccount: pod-cpu-hog-sa
  monitoring: false
  components:
    runner:
      runnerannotation:
        sidecar.istio.io/inject: "false"
  # It can be delete/retain
  jobCleanUpPolicy: 'delete'
  experiments:
    - name: pod-cpu-hog
      spec:
        components:
          experimentannotation:
            sidecar.istio.io/inject: "false"
          env:
            #number of cpu cores to be consumed
            #verify the resources the app has been launched with
            - name: CPU_CORES
              value: '1'

            - name: TOTAL_CHAOS_DURATION
              value: '60' # in seconds

**Comment by arunsriniv** [comment]
```
apiVersion: litmuschaos.io/v1alpha1
kind: ChaosEngine
metadata:
  name: avc-platform-gateway-chaos
  namespace: avc
spec:
  # It can be true/false
  annotationCheck: 'false'
  # It can be active/stop
  engineState: 'active'
  appinfo:
    appns: 'avc'
    applabel: 'app=avc-platform-gateway'
    appkind: 'deployment'
  chaosServiceAccount: pod-cpu-hog-sa
  monitoring: false
  components:
    runner:
      runnerannotation:
        sidecar.istio.io/inject: "false"
  # It can be delete/retain
  jobCleanUpPolicy: 'delete'
  experiments:
    - name: pod-cpu-hog
      spec:
        components:
          experimentannotation:
            sidecar.istio.io/inject: "false"
          env:
            #number of cpu cores to be consumed
            #verify the resources the app has been launched with
            - name: CPU_CORES
              value: '1'

            - name: TOTAL_CHAOS_DURATION
              value: '60' # in seconds

```

**Comment by ispeakc0de** [comment]
Thanks for sharing the manifest. It doesn't require providing the `LIB_IMAGE_PULL_POLICY`  explicitly as an env inside chaos-experiment. It will automatically use the imagePullPolicy provided at `spec.definition.imagePullPolicy` .

Can you please share the ouput of following command, just want to see if imagePullPolicy is set there:

```yaml
kubectl describe chaosexperiment pod-cpu-hog -n avc
```

**Comment by arunsriniv** [comment]
Thanks @ispeakc0de . I dont see imagePullPolicy set when I do describe. Here is what I get.

`-bash-4.2$ kubectl -n avc describe chaosexperiment node-cpu-hog
Name:         node-cpu-hog
Namespace:    avc
Labels:       app.kubernetes.io/component=chaosexperiment
              app.kubernetes.io/part-of=litmus
              app.kubernetes.io/version=1.13.2
              name=node-cpu-hog
Annotations:  API Version:  litmuschaos.io/v1alpha1
Description:
  Message:  Give a cpu spike on a node belonging to a deployment

Kind:  ChaosExperiment
Metadata:
  Creation Timestamp:  2021-05-13T11:02:02Z
  Generation:          3
  Managed Fields:
    API Version:  litmuschaos.io/v1alpha1
    Fields Type:  FieldsV1
    fieldsV1:
      f:description:
        .:
        f:message:
      f:metadata:
        f:annotations:
          .:
          f:kubectl.kubernetes.io/last-applied-configuration:
        f:labels:
          .:
          f:app.kubernetes.io/component:
          f:app.kubernetes.io/part-of:
          f:app.kubernetes.io/version:
          f:name:
      f:spec:
        .:
        f:definition:
          .:
          f:args:
          f:command:
          f:env:
          f:image:
          f:labels:
            .:
            f:app.kubernetes.io/component:
            f:app.kubernetes.io/part-of:
            f:app.kubernetes.io/version:
            f:name:
          f:permissions:
          f:scope:
    Manager:         kubectl
    Operation:       Update
    Time:            2021-05-13T11:07:39Z
  Resource Version:  221936
  Self Link:         /apis/litmuschaos.io/v1alpha1/namespaces/avc/chaosexperiments/node-cpu-hog
  UID:               11ac6164-1bf3-495d-9114-7fb3b938e734
Spec:
  Definition:
    Args:
      -c
      ./experiments -name node-cpu-hog
    Command:
      /bin/bash
    Env:
      Name:   TOTAL_CHAOS_DURATION
      Value:  60
      Name:   RAMP_TIME
      Value:
      Name:   NODE_CPU_CORE
      Value:
      Name:   TARGET_NODES
      Value:
      Name:   LIB
      Value:  litmus
      Name:   LIB_IMAGE
      Value:  litmuschaos/go-runner:1.13.2
      Name:   NODES_AFFECTED_PERC
      Value:
      Name:   SEQUENCE
      Value:  parallel
    Image:    litmuschaos/go-runner:1.13.2
    Labels:
      app.kubernetes.io/component:  experiment-job
      app.kubernetes.io/part-of:    litmus
      app.kubernetes.io/version:    1.13.2
      Name:                         node-cpu-hog
    Permissions:
      API Groups:

        batch
        litmuschaos.io
      Resources:
        jobs
        pods
        pods/log
        pods/exec
        events
        chaosengines
        chaosexperiments
        chaosresults
      Verbs:
        create
        list
        get
        patch
        update
        delete
        deletecollection
      API Groups:

      Resources:
        nodes
      Verbs:
        get
        list
    Scope:  Cluster
Events:     <none>
`

**Comment by arunsriniv** [comment]
Above is for node-cpu-hog, same observed for other experiments as well (including pod-cpu-hog, container-kill etc). Basically, the `imagePullPolicy` is taken into consideration and it is taking the default which is `Always`

**Comment by ispeakc0de** [comment]
Yes. I am suspecting that you have older CRDs. Can you please update the CRDs with the following command:
```yaml
kubectl apply -f https://raw.githubusercontent.com/litmuschaos/chaos-operator/master/deploy/chaos_crds.yaml
```

after this re-apply the experiment CR as well

**Comment by arunsriniv** [comment]
Excellent. Yes, it works now.  Thanks a lot @ispeakc0de 

```
    Image:              litmuschaos/go-runner:1.13.2
    Image Pull Policy:  IfNotPresent
```

**Pull request merged**